### PR TITLE
Adding an integration test

### PR DESCRIPTION
**Pull request by vishh**
The intent is to run this periodically (nightly maybe).

### CODE REVIEW

**Comment by vishh** [comment]
Ping @rjnagal @vmarmol

**Comment by vmarmol** [comment]
Should we remove the integration.test file?

File: integration/framework.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+func exists(path string) bool {
+	if _, err := os.Stat(path); err != nil {
+		return false
+	}
+	return true
+}
+
+const pathToGCEConfig = "cluster/gce/config-default.sh"
+
+func readFile(file string) (string, error) {
```
Comment (vmarmol) [inline integration/framework.go]: Why have this function over just ioutil.ReadFile?

**Comment by vishh** [comment]
Wonder how it sneaked in. Removed.

File: integration/framework.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+	for i, line := range lines {
+		if strings.Contains(line, "ENABLE_CLUSTER_MONITORING") {
+			lines[i] = "ENABLE_CLUSTER_MONITORING=false"
+		} else if strings.Contains(line, "NUM_MINIONS=") {
+			lines[i] = "NUM_MINIONS=2"
+		}
+	}
+	output := strings.Join(lines, "\n")
+	err = ioutil.WriteFile(kubeConfigFilePath, []byte(output), 0644)
```
Comment (vmarmol) [inline integration/framework.go]: nit: return ioutil...

File: integration/framework.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+func readFile(file string) (string, error) {
+	input, err := ioutil.ReadFile(file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(input), nil
+}
+
+func disableClusterMonitoring(kubeBaseDir string) error {
```
Comment (vmarmol) [inline integration/framework.go]: maybe: changeClusterConfig() and document what it changes?

**Comment by vishh** [comment]
That function doesn't serve any purpose now. Removed.

File: integration/framework.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+	if err := disableClusterMonitoring(kubeBaseDir); err != nil {
+		return nil, fmt.Errorf("failed to disable cluster monitoring in kube cluster config - %q", err)
+	}
+	glog.V(1).Info("Disabled cluster monitoring")
+
+	if !*useExistingCluster {
+		// TODO(vishh): Do a kube-push if a cluster already exists and bring down monitoring jobs.
+		// Teardown any pre-existing cluster - This can happen due to a failed test or bad release.
+		// Ignore teardown errors since a cluster may not even exist.
+		_ = tearDownCluster(kubeBaseDir)
```
Comment (vmarmol) [inline integration/framework.go]: Do we need the "_ ="?

File: integration/framework.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+	if testing.Short() {
+		t.Skip("Skipping framework test in short mode")
+	}
+
+	if !exists(*authConfig) {
+		return nil, fmt.Errorf("failed to find kube auth file: %s", *authConfig)
+	}
+
+	// Create a temp dir to store the kube release files.
+	tempDir := path.Join(*workDir, version)
```
Comment (vmarmol) [inline integration/framework.go]: Consider using [TempDir](http://golang.org/pkg/io/ioutil/#TempDir).

File: integration/framework.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+func downloadRelease(workDir, version string) error {
+	// Temporary download path.
+	downloadPath := path.Join(workDir, "kube")
+	// Format url.
+	downloadUrl := fmt.Sprintf(imageUrlTemplate, version)
+	glog.V(1).Infof("About to download kube release using url: %q", downloadUrl)
+
+	// Download kube code and store it in a temp dir.
+	if err := exec.Command("wget", downloadUrl, "-O", downloadPath).Run(); err != nil {
+		return fmt.Errorf("failed to wget kubernetes release @ %s - %q", downloadUrl, err)
```
Comment (vmarmol) [inline integration/framework.go]: %q and %v instead. Same below

File: integration/framework.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	}
+	glog.V(2).Infof("Output of gcloud compute instance list kubernetes-master:\n %s", out)
+	for _, line := range strings.Split(string(out), "\n") {
+		if strings.Contains(line, "natIP") {
+			externalIPLine := strings.Split(line, ":")
+			return strings.TrimSpace(externalIPLine[1]), nil
+		}
+	}
+
+	return "", fmt.Errorf("External IP not found for the master. 'gcloud compute instances list kubernetes-master' returned %s", out)
```
Comment (vmarmol) [inline integration/framework.go]: nit: s/External/external/

File: integration/framework.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+	}
+	return nil
+}
+
+func runKubeClusterCommand(kubeBaseDir, command string) ([]byte, error) {
+	return exec.Command(path.Join(kubeBaseDir, "cluster", command)).CombinedOutput()
+}
+
+func setupNewCluster(kubeBaseDir string) error {
+	out, err := exec.Command(path.Join(kubeBaseDir, "cluster", "kube-up.sh")).CombinedOutput()
```
Comment (vmarmol) [inline integration/framework.go]: runKubeClusterCommand()? Here and in tearDownCluster

File: integration/kube_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		if _, rerr := rcInterface.Update(controller); rerr != nil {
+			return rerr
+		}
+	}
+	glog.Errorf("updateReplicas failed for controller %+v with error: %q", controller, err)
+	return err
+}
+
+func deletePods(fm kubeFramework) {
+	// TODO(vishh): Fix this.
```
Comment (vmarmol) [inline integration/kube_test.go]: Maybe more comment on what "this" is? :)

File: integration/kube_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	if err := updateReplicas(fm, heapsterController, 0); err != nil {
+		glog.Errorf("failed to bring down the number of replicas for heapster")
+	}
+	_, _ = fm.RunKubectlCmd("delete", "-f", *heapsterManifestFile)
+	_, _ = fm.RunKubectlCmd("delete", "-f", *influxdbGrafanaManifestFile)
+	_, _ = fm.RunKubectlCmd("delete", "-f", *influxdbServiceFile)
+}
+
+func TestHeapsterInfluxDBWorks(t *testing.T) {
```
Comment (vmarmol) [inline integration/kube_test.go]: consider doing:

require = require.New(t)

which lets you no longer have to specify t in the function

File: integration/framework.go (hunk 1 above)
Comment (vishh) [inline integration/framework.go]: Removed.

File: integration/framework.go (hunk 2 above)
Comment (vishh) [inline integration/framework.go]: Done

File: integration/framework.go (hunk 4 above)
Comment (vishh) [inline integration/framework.go]: no :) removed.

File: integration/framework.go (hunk 5 above)
Comment (vishh) [inline integration/framework.go]: The intent is to reuse kube releases across runs. That is the reason for using the same tempdir.

File: integration/framework.go (hunk 6 above)
Comment (vishh) [inline integration/framework.go]: Done

File: integration/framework.go (hunk 7 above)
Comment (vishh) [inline integration/framework.go]: Done.

File: integration/framework.go (hunk 8 above)
Comment (vishh) [inline integration/framework.go]: Done

File: integration/kube_test.go (hunk 9 above)
Comment (vishh) [inline integration/kube_test.go]: Done

File: integration/kube_test.go (hunk 10 above)
Comment (vishh) [inline integration/kube_test.go]: Done.

**Comment by vishh** [comment]
@vmarmol addressed your comments. Don't merge it just yet. I want to test it once again.

File: integration/framework.go (hunk 5 above)
Comment (vmarmol) [inline integration/framework.go]: Ah kk. WDYT of using TempDir to generate workDir? We can have the flag's default value be empty which means "make one for me"

**Comment by vmarmol** [comment]
LGTM

**Comment by vishh** [comment]
This is ready to be merged.

**Comment by vishh** [comment]
Self merging this.

**Pull request merged**